README.md
<img src="https://github.com/AraBlocks/docs/blob/master/ara.png" width="30" height="30" /> ara-crypto
==========

Cryptographic functions used in ARA modules for Node.js.

## Status

This project is in active development.

## Stability

> [Stability][stability-index]: 1 - Experimental. This feature is still under
> active development and subject to non-backwards compatible changes, or even
> removal, in any future version. Use of the feature is not recommended
> in production environments. Experimental features are not subject to
> the Node.js Semantic Versioning model.

## Installation

```sh
$ npm install arablocks/ara-crypto
```

## Usage

```js
const crypto = require('ara-crypto')

const message = Buffer.from('message')
const bytes = crypto.randomBytes(32)
const hash = crypto.blake2b(message)
const { publicKey, secretKey } = crypto.keyPair(hash) // hash seed optional
const signature = crypto.sign(message, secretKey)
const verified = crypto.verify(signature, message, publicKey)
const buffer = crypto.uint64.encode(80)
const number = crypto.uint64.decode(buffer) // 80
```

## API

Most of the functions exported by this module will check for input
correctness. If given incorrect input, a function will throw a
`TypeError` with a message describing the error. In most functions,
inputs should always be a [`Buffer`](https://nodejs.org/api/buffer.html).

- [crypto.randomBytes(size)](#randomBytes)
- [crypto.blake2b(data)](#blake2b)
- [crypto.discoveryKey(\[seed\]\[, size\]\[, key\])](#discoveryKey)
- [crypto.keyPair(\[seed\])](#keyPair)
- [crypto.sign(message, secretKey)](#sign)
- [crypto.verify(signature, message, publicKey)](#verify)
- [crypto.curve25519.keyPair(\[seed\])](#curve25519-keyPair)
- [crypto.curve25519.shared(secretKey, publicKey)](#curve25519-shared)
- [crypto.ed25519.keyPair(\[seed\])](#ed25519-keyPair)
- [crypto.ed25519.sign(message, secretKey)](#ed25519-sign)
- [crypto.ed25519.verify(signature, message, publicKey)](#ed25519-verify)
- [crypto.uint64.encode(number)](#uint64-encode)
- [crypto.uint64.decode(buffer)](#uint64-decode)
- [crypto.box(buffer, opts)](#box)
- [crypto.unbox(buffer, opts)](#unbox)
- [crypto.createBoxStream(opts)](#createBoxStream)
- [crypto.createUnboxStream(opts)](#createUnboxStream)
- [crypto.auth(message, key)](#auth)
- [crypto.auth.verify(mac, message, key)](#auth-verify)
- [crypto.kx.keyPair(\[seed\])](#kx-keyPair)
- [crypto.kx.client(opts)](#kx-client)
- [crypto.kx.server(opts)](#kx-server)
- [crypto.seal(message, opts)](#seal)
- [crypto.seal.open(message, opts)](#seal-open)
- [crypto.shash(message, secretKey)](#shash)

### `crypto.randomBytes(size)` <a name="randomBytes"></a>

> **Stability: 2** - Stable

Generate a buffer of random bytes where `size` is an unsigned integer
greater than `0`. This function will throw a `TypeError` if given
incorrect input. This function calls `sodium.randombytes_buf`
internally.

```js
const bytes = crypto.randomBytes(32)
```

### `crypto.blake2b(buffer[, size])` <a name="blake2b"></a>

> **Stability: 2** - Stable

Generates a blake2b digest hash from input of a
given size defaulting to 32 bytes. This function calls
`crypto_generichash_batch` internally.

```js
const hash = crypto.blake2b(Buffer.from("message"))
```

### `crypto.discoveryKey([seed][, size][, key])` <a name="discoveryKey"></a>

> **Stability: 2** - Stable

Generate a discovery key useful for network
keys. This function calls `crypto_generichash` internally.

```js
const { publicKey, secretKey } = crypto.keyPair()
const discoveryKey = crypto.discoveryKey(publicKey)
```

### `crypto.keyPair([seed])` <a name="keyPair"></a>

An alias for [crypto.ed25519.keyPair](#ed25519-keyPair).

### `crypto.sign(message, secretKey)` <a name="sign"></a>

An alias for [crypto.ed25519.sign](#ed25519-sign).

### `crypto.verify(signature, message, publicKey)` <a name="verify"></a>

An alias for [crypto.ed25519.verify](#ed25519-verify).

### `crypto.curve25519.keyPair([seed])` <a name="curve25519-keyPair"></a>

> **Stability: 2** - Stable

Generate a Curve25519 public and secret key pair from an optional
seed buffer. This function calls `crypto_sign_seed_keypair` and
`crypto_sign_keypair` internally and converts to Curve25519 key pair
calling `crypto_sign_ed25519_pk_to_curve25519` and
`crypto_sign_ed25519_sk_to_curve25519`.

```js
const seed = crypto.randomBytes(32)
const { publicKey, secretKey } = crypto.curve25519.keyPair(seed)
```

### `crypto.curve25519.shared(secretKey, publicKey)`

Compute a shared key from a 32 byte secret key and a 32 byte public
key. If keys are larger than 32 bytes, they will be truncated when read.

```js
const alice = crypto.curve25519.keyPair()
const bob = crypto.curve25519.keyPair()
const shared = crypto.curve25519.shared(alice.secretKey, bob.publicKey)
```

### `crypto.ed25519.keyPair([seed])` <a name="ed25519-keyPair"></a>

> **Stability: 2** - Stable

Generate a public and secret key pair from an optional
seed buffer. This function will throw a `TypeError` if given incorrect input.
This function calls `crypto_sign_seed_keypair` and `crypto_sign_keypair`
internally.

```js
const seed = crypto.randomBytes(32)
const { publicKey, secretKey } = crypto.keyPair(seed)
```

### `crypto.ed25519.sign(message, secretKey)` <a name="ed25519-sign"></a>

> **Stability: 2** - Stable

Sign a message buffer with a secret key buffer. This function will throw
a `TypeError` if given incorrect input. This function calls
`crypto_sign_detached` on a buffer of size `crypto_sign_BYTES`.

```js
const { publicKey, secretKey } = crypto.keyPair()
const signature = crypto.sign(Buffer.from("hello"), secretKey)
```

### `crypto.ed25519.verify(signature, message, publicKey)` <a name="ed25519-verify"></a>

> **Stability: 2** - Stable

Verify signature for a message signed with a given
public key. This function will throw a `TypeError` if given incorrect
input. This function calls `crypto_sign_verify_detached`
internally.

```js
const { publicKey, secretKey } = crypto.keyPair()
const message = Buffer.from("hello")
const signature = crypto.sign(message, secretKey)
const verified = crypto.verify(signature, message, publicKey)
if (verified) {
  // message was signed with secret key (corresponding
  // to the given public key) that generated the given signature
}
```

### `crypto.uint64.encode(value, size)` <a name="uint64-encode"></a>

> **Stability: 2** - Stable

Encode an unsigned 64-bit big endian number into a buffer
of a given size defaulting to 8 bytes.

```js
const buffer = crypto.uint64.encode(80)
```

### `crypto.uint64.decode(buffer)` <a name="uint64-decode"></a>

> **Stability: 2** - Stable

Decode an unsigned 64-bit big endian buffer into a number

```js
const buffer = crypto.uint64.encode(80)
const number = crypto.uint64.decode(buffer) // 80
```

### `crypto.box(buffer, opts)` <a name="box"></a>

> **Stability: 1** - Experimental

"Boxes", or encrypts, a buffer from a 32 byte encryption key and a 24-byte nonce.

```js
const key = Buffer.alloc(32); key.fill('SECRET!KEY')
const nonce = crypto.randomBytes(24)
const secret = Buffer.concat([ key, nonce ])
const buffer = Buffer.from('hello!')
const boxed = crypto.box(buffer, { secret }) // or crypto.box(buffer, { nonce, key })
console.log(boxed) // <Buffer 11 8f 40 2b 8a f5 10 08 1f fe 59 b9 97 9c b8 a2 89 e7 b8 78 50 75 ed d9 8e 9c 09 38 0e 81 31 ff fa c6 96 df 57 db 85 ae>

```

### `crypto.unbox(buffer, opts)` <a name="unbox"></a>

> **Stability: 1** - Experimental

"Unboxes" or decrypts a buffer from a 32-byte encryption key and a 24-byte nonce.

```js
const key = Buffer.alloc(32); key.fill('SECRET!KEY')
const nonce = crypto.randomBytes(24)
const secret = Buffer.concat([ key, nonce ])
const buffer = Buffer.from('hello!')
const boxed = crypto.box(buffer, { secret }) // or crypto.box(buffer, { nonce, key })
const unboxed = crypto.unbox(boxed, { secret }) // or crypto.unbox(boxed, { nonce, key })
console.log(unboxed) // hello!
```

### `crypto.createBoxStream(opts)` <a name="createBoxStream"></a>

> **Stability: 1** - Experimental

Creates a transform stream that "boxes" messages written to it.

```js
const key = Buffer.alloc(32); key.fill('SECRET!KEY')
const nonce = crypto.randomBytes(24)
const secret = Buffer.concat([ key, nonce ])
const buffer = Buffer.from('hello!')
const stream = crypto.createBoxStream({ secret }) // or crypto.createBoxStream({ nonce, key })
stream.on('data', (chunk) => console.log(chunk)) // cipher text
stream.write(buffer)
```

### `crypto.createUnboxStream(opts)` <a name="createUnboxStream"></a>

> **Stability: 1** - Experimental

Creates a transform stream that "unboxes" messages written to it.

```js
const key = Buffer.alloc(32); key.fill('SECRET!KEY')
const nonce = crypto.randomBytes(24)
const secret = Buffer.concat([ key, nonce ])
const buffer = Buffer.from('hello!')
const stream = crypto.createUnboxStream({ secret }) // or crypto.createUnboxStream({ nonce, key })
const boxed = crypto.box(buffer, { secret })
stream.on('data', (chunk) => console.log(chunk)) // hello!
stream.write(boxed)
```

### `crypto.auth(message, key)` <a name="auth"></a>

> **Stability: 2** - Stable

Generates and returns a message authentication code (MAC) for
a given message and secret key.

```js
const message = Buffer.from('hello')
const key = crypto.randomBytes(32)
const mac = auth(message, key)
```

`Secret + Nonce` based message authentication codes:

```js
const message = Buffer.from('hello')
const secret = getSecretFromSomeWhere()
const nonce = crypto.randomBytes(32)
const key = crypto.blake2b(Buffer.concat([secret, nonce]))
const mac = auth(message, key)
```

### `crypto.auth.verify(mac, message, key)` <a name="auth-verify"></a>

> **Stability: 2** - Stable

Verifies the authenticity of a message with a given message
authentication code (MAC) and secret key.

```js
const message = Buffer.from('hello')
const key = crypto.randomBytes(32)
const mac = auth(message, key)

if (false === verify(mac, message, key)) {
  throw new Error('Message forged!')
}
```

### `crypto.kx.keyPair([seed])` <a name="kx-keyPair"></a>

> **Stability: 2** - Stable

Generates a key exchange key pair.

```js
const seed = crypto.randomBytes(32)
const kp = crypto.kx.keyPair(seed)
```

### `crypto.kx.client(opts)` <a name="kx-client"></a>

> **Stability: 1** - Experimental

Compute sender (tx) and receiver (rx) session keys for a client
based on a server's public key.

```js
const serverPublicKey = getServerPublicKey()
const { publicKey, secretKey } = getClientKeyPair()
const client = kx.client({
  publicKey,
  secretKey,
  server: { publicKey: serverPublicKey }
})
```

### `crypto.kx.server(opts)` <a name="kx-server"></a>

> **Stability: 1** - Experimental

Compute sender (tx) and receiver (rx) session keys for a server
based on a client's public key.

```js
const clientPublicKey = getClientPublicKey()
const { publicKey, secretKey } = getServerKeyPair()
const server = kx.server({
  publicKey,
  secretKey,
  client: { publicKey: clientPublicKey }
})
```

### `crypto.kdf.keygen(key)` <a name="kdf-keygen"></a>

> **Stability: 2** - Stable

Generate a master ("secret") key. This function calls `crypto_kdf_keygen`
internally.

```js
const key = crypto.kdf.keygen()
```

### `crypto.kdf.init(key, buffer)` <a name="kdf-init"></a>

> **Stability: 2** - Stable

Initializes key and buffer with null subkey to return an object to update.

```js
const buffer = Buffer.from('examples')
const key = crypto.kdf.keygen()
const ctx = crypto.kdf.init(key, buffer) // buffer optional
```

### `crypto.kdf.update(ctx, id)` <a name="kdf-update"></a>

> **Stability: 2** - Stable

Updates context subkey from an ID. This function calls `crypto_kdf_derive`
internally.

```js
const buffer = Buffer.from('examples')
const key = crypto.kdf.keygen()
const ctx = crypto.kdf.init(key, buffer)
const subkey = crypto.kdf.update(ctx, 1)
```

### `crypto.kdf.final(ctx)` <a name="kdf-final"></a>

> **Stability: 2** - Stable

Finalizes context by setting `ctx.subkey` to `null`.

```js
const buffer = Buffer.from('examples')
const key = crypto.kdf.keygen()
const ctx = crypto.kdf.init(key, buffer)
const subkey = crypto.kdf.final(ctx)
```

### `crypto.kdf.derive(key, iterations, buffer)` <a name="kdf-derive"></a>

> **Stability: 2** - Stable

Derives a subkey from a key, number of iterations, and a context buffer.
This function calls `kdf.init`, `kdf.update`, and `kdf.final` internally.

```js
const buffer = Buffer.from('examples')
const key = crypto.kdf.keygen()
const subkey = crypto.kdf.derive(key, 1, buffer) // buffer optional
```

### `crypto.seal(message, opts)` <a name="seal"></a>

> **Stability: 2** - Stable

Seals a message based on a curve25519 public key for a recipient
who has the corresponding secret key.

```js
const publicKey = getServerPublicKey()
const message = Buffer.from('hello')
const sealed = crypto.seal(message, { publicKey })
```

### `crypto.seal.open(message, opts)` <a name="seal-open"></a>

> **Stability: 2** - Stable

Opens a sealed message based on a curve25519 public key for a recipient
who has the corresponding secret key.

```js
const { publicKey, secretKey } = getKeyPair()
const message = crypto.seal.open(ciphertext, { publicKey, secretKey })
```

### `crypto.shash(message, secretKey)` <a name="shash"></a>

> **Stability: 2** - Stable

Compute a 8 byte short hash for some message buffer based on
a given secret key.

```js
const message = Buffer.from('message')
const keyPair = crypto.curve25519.keyPair()
const key = crypto.blake2b(keyPair.secretKey, 16)
const mac = crypto.shash(message, key)
```

## Contributing

- [Commit message format](/.github/COMMIT_FORMAT.md)
- [Commit message examples](/.github/COMMIT_FORMAT_EXAMPLES.md)
- [How to contribute](/.github/CONTRIBUTING.md)

## See Also

- [libsodium](https://download.libsodium.org/doc/)
- [sodium-universal](https://github.com/sodium-friends/sodium-universal)
- [ara-identity](https://github.com/arablocks/ara-identity)
- [Key derivation function](https://en.wikipedia.org/wiki/Key_derivation_function)
- [Stability index][stability-index]

## License

LGPL-3.0

[stability-index]: https://nodejs.org/api/documentation.html#documentation_stability_index


CHANGELOG.md
## [0.9.4](https://github.com/AraBlocks/ara-crypto/compare/0.9.3...0.9.4) (2021-04-02)



## [0.9.3](https://github.com/AraBlocks/ara-crypto/compare/0.9.2...0.9.3) (2021-04-02)



## [0.9.2](https://github.com/AraBlocks/ara-crypto/compare/0.9.1...0.9.2) (2021-04-02)


### Bug Fixes

* unbox nonce and key length ([033d294](https://github.com/AraBlocks/ara-crypto/commit/033d2944e50e2d9e10d6157bd851da5812c946b7))



## [0.9.1](https://github.com/AraBlocks/ara-crypto/compare/0.9.0...0.9.1) (2021-04-02)


### Bug Fixes

* **kdf.js:** truncate key to exact bytes needed ([e502c6b](https://github.com/AraBlocks/ara-crypto/commit/e502c6befd4ed4bd8a3505c22feff351f3d3e833))



# [0.9.0](https://github.com/AraBlocks/ara-crypto/compare/0.8.5...0.9.0) (2021-04-01)



## [0.8.5](https://github.com/AraBlocks/ara-crypto/compare/0.8.4...0.8.5) (2019-04-05)



## [0.8.4](https://github.com/AraBlocks/ara-crypto/compare/0.8.3...0.8.4) (2019-03-27)


### Bug Fixes

* **package.json:** Lock in sodium-browserify@1.2.6 ([b389138](https://github.com/AraBlocks/ara-crypto/commit/b389138b345e26679f11f6606b2db6ff0e51345c))



## [0.8.3](https://github.com/AraBlocks/ara-crypto/compare/0.8.1...0.8.3) (2019-03-27)


### Bug Fixes

* **kdf:** Remove key length constraint. ([2f3103f](https://github.com/AraBlocks/ara-crypto/commit/2f3103f58cd9cf905e620511a865e3aab6fc263a))



## [0.8.1](https://github.com/AraBlocks/ara-crypto/compare/0.8.0...0.8.1) (2018-11-09)



# [0.8.0](https://github.com/AraBlocks/ara-crypto/compare/0.7.1...0.8.0) (2018-11-09)


### Bug Fixes

* **kdf:** Alloc context buffer in init instead of allocUnsafe ([7c6d261](https://github.com/AraBlocks/ara-crypto/commit/7c6d2612fbcec7a380e9366bd20d3a6ce6dd9d51))
* **kdf:** Make buffer optional for 'init' and 'derive'. ([99d46f1](https://github.com/AraBlocks/ara-crypto/commit/99d46f1f4ea7ba4e20204aa06887b96b740478ae))
* **kdf:** Remove unused constants crypto_kdf_BYTES_M* ([b485c9c](https://github.com/AraBlocks/ara-crypto/commit/b485c9c183e231d540a4fa708381686a3d9738df))
* **kdf.js:** Destructure value in ctx object. ([8dffd65](https://github.com/AraBlocks/ara-crypto/commit/8dffd65422da463cce38cdd5208891d8b61ed5d2))
* **kdf.js:** Fix eslint formatting ([a912996](https://github.com/AraBlocks/ara-crypto/commit/a9129967c569471febf0cb0c3ca9d2d883a1ebd1))
* **kdf.js:** Update at i + 1 ([fe92c45](https://github.com/AraBlocks/ara-crypto/commit/fe92c450ed3ae67bf7ec1ec9905349b9b9f2ef00))


### Features

* **kdf.js:** Introduce key derivation functions ([b7d8bf0](https://github.com/AraBlocks/ara-crypto/commit/b7d8bf03dc5ec2c302930f017778ff397e0e8f92))



## [0.7.1](https://github.com/AraBlocks/ara-crypto/compare/0.7.0...0.7.1) (2018-10-17)



# [0.7.0](https://github.com/AraBlocks/ara-crypto/compare/0.6.0...0.7.0) (2018-08-14)


### Features

* **test/helpers:** Introduce test runner and browser index ([d1a95b6](https://github.com/AraBlocks/ara-crypto/commit/d1a95b6043c584631f911f7bebbf288708ad2690))



# [0.6.0](https://github.com/AraBlocks/ara-crypto/compare/0.5.1...0.6.0) (2018-08-14)



## [0.5.1](https://github.com/AraBlocks/ara-crypto/compare/0.5.0...0.5.1) (2018-08-09)



# [0.5.0](https://github.com/AraBlocks/ara-crypto/compare/0.4.2...0.5.0) (2018-08-06)



## [0.4.2](https://github.com/AraBlocks/ara-crypto/compare/0.4.1...0.4.2) (2018-08-01)


### Features

* **shash.js:** Introduce short hash ([b8bc6f1](https://github.com/AraBlocks/ara-crypto/commit/b8bc6f10fab38ef2ae77e970d63b5f5ac0e03938))



## [0.4.1](https://github.com/AraBlocks/ara-crypto/compare/0.4.0...0.4.1) (2018-07-31)


### Features

* **curve25519.js:** Introduce shared keys (scalar multiplcation) ([8497317](https://github.com/AraBlocks/ara-crypto/commit/84973176e35d269c08ef360f2cf895717edf0d5d))



# [0.4.0](https://github.com/AraBlocks/ara-crypto/compare/0.3.2...0.4.0) (2018-07-30)


### Features

* **auth.js:** Introduce message authentication code API ([1c1d197](https://github.com/AraBlocks/ara-crypto/commit/1c1d1976d193929eac01cc148db6258a6b028708))
* **box.js:** Introduce box encryption ([f4dae43](https://github.com/AraBlocks/ara-crypto/commit/f4dae4339872ec512a9546d6c76f6accab3413fd))
* **ed25519.js:** Move ed25519 keyPair/sign/auth functions to scoped module ([8a86697](https://github.com/AraBlocks/ara-crypto/commit/8a8669718f7b34a48c674c6f9776262ebcdae468))
* **kx.js:** Introduce key exchange API ([fffafe6](https://github.com/AraBlocks/ara-crypto/commit/fffafe6bfd738271b9954c1f8e93fb3e9515ec43))
* **seal.js:** Introduce sealed box API ([13e2219](https://github.com/AraBlocks/ara-crypto/commit/13e2219cc0e264b75ecd857bda8fc2210ea5b02c))
* **unbox.js:** Introduce box decryption ([c31cd7b](https://github.com/AraBlocks/ara-crypto/commit/c31cd7b3d8e10cb85d7be07dba6206e3a59207aa))



## [0.3.2](https://github.com/AraBlocks/ara-crypto/compare/0.3.1...0.3.2) (2018-06-19)


### Features

* **curve25519.js:** Introduce Curve25519 key pairs ([5d0e731](https://github.com/AraBlocks/ara-crypto/commit/5d0e73175b53d86edaf53ea5ba36b8c8eef8cc54))



## [0.3.1](https://github.com/AraBlocks/ara-crypto/compare/0.3.0...0.3.1) (2018-05-30)



# [0.3.0](https://github.com/AraBlocks/ara-crypto/compare/0.2.4...0.3.0) (2018-05-30)


### Features

* **base58.js:** Introduce simple base58 wrapper ([a21649b](https://github.com/AraBlocks/ara-crypto/commit/a21649bcc85937230a688b3232d4d8cd48937058))
* **base64.js:** Introduce simple base64 wrapper ([ab5d54d](https://github.com/AraBlocks/ara-crypto/commit/ab5d54d47b84382253bc562665ad0cfee4a52115))



## [0.2.4](https://github.com/AraBlocks/ara-crypto/compare/0.2.3...0.2.4) (2018-05-27)



## [0.2.3](https://github.com/AraBlocks/ara-crypto/compare/0.2.2...0.2.3) (2018-05-24)



## [0.2.2](https://github.com/AraBlocks/ara-crypto/compare/0.2.1...0.2.2) (2018-05-22)



## [0.2.1](https://github.com/AraBlocks/ara-crypto/compare/0.2.0...0.2.1) (2018-05-22)


### Features


box.js
const increment = require('increment-buffer')
const isBuffer = require('is-buffer')
const through = require('through2')
const split = require('split-buffer')

/* eslint-disable camelcase */
const {
  crypto_secretbox_NONCEBYTES,
  crypto_secretbox_KEYBYTES,
  crypto_secretbox_MACBYTES,

  crypto_secretbox_easy,
} = require('./sodium')

const kBoxHeaderSize = 2 + (2 * crypto_secretbox_MACBYTES)
const kBoxBufferSize = 4 * 1024

/**
 * "Boxes", or encrypts, a buffer from a 32 byte encryption key
 * and a 24-byte nonce.
 *
 * @public
 * @param {Buffer} buffer
 * @param {Object} opts
 * @param {Buffer} opts.buffer
 * @param {?(Buffer)} opts.secret
 * @param {?(Buffer)} opts.nonce
 * @param {?(Buffer)} opts.key
 * @return {Buffer}
 * @throws TypeError
 */
function box(buffer, opts) {
  if (!opts || 'object' !== typeof opts) {
    throw new TypeError('crypto.box: Expecting object.')
  }

  const { secret } = opts
  let { nonce, key } = opts

  if (!buffer || false === isBuffer(buffer)) {
    throw new TypeError('crypto.box: Expecting buffer.')
  }

  if (isBuffer(secret) && secret.length >= crypto_secretbox_KEYBYTES) {
    key = secret.slice(0, crypto_secretbox_KEYBYTES)
    nonce = isBuffer(opts.nonce)
      ? opts.nonce
      : secret.slice(crypto_secretbox_KEYBYTES)
  }

  if (false === isBuffer(nonce)) {
    throw new TypeError('crypto.box: Expecting nonce.')
  }

  if (false === isBuffer(key)) {
    throw new TypeError('crypto.box: Expecting secret key.')
  }

  // truncate key and nonce
  nonce = nonce.slice(0, crypto_secretbox_NONCEBYTES)
  key = key.slice(0, crypto_secretbox_KEYBYTES)

  // ephemeral nonces used for header and body buffer boxing
  const nonces = [ copy(nonce), increment(copy(nonce)) ]

  // length(2) + MAC(16) == 18
  const header = Buffer.alloc(2 + crypto_secretbox_MACBYTES)

  // boxed buffer cipher text
  const body = Buffer.alloc(crypto_secretbox_MACBYTES + buffer.length)

  // output buffer frames
  const frames = [
    Buffer.alloc(crypto_secretbox_MACBYTES + header.length),
    Buffer.alloc(buffer.length),
  ]

  // write buffer length into header
  header.writeUInt16BE(buffer.length, 0)

  // box message buffer
  crypto_secretbox_easy(body, buffer, nonces[1], key)

  // copy MAC into header after length (offset 2)
  body.copy(
    header,
    2,
    0,
    crypto_secretbox_MACBYTES
  )

  // copy boxed message buffer into frames[1]
  body.copy(
    frames[1],
    0,
    crypto_secretbox_MACBYTES,
    crypto_secretbox_MACBYTES + buffer.length
  )

  // box header buffer into frame[0] based on current nonces[0] and key
  crypto_secretbox_easy(frames[0], header, nonces[0], key)

  return Object.assign(Buffer.concat(frames), { nonce: nonces[1] })
}

/**
 * Creates a transform stream that "boxes" messages
 * written to it.
 *
 * @public
 * @param {Object} opts
 * @param {?(Buffer)} opts.secret
 * @param {?(Buffer)} opts.nonce
 * @param {?(Buffer)} opts.key
 * @return {Stream}
 * @throws TypeError
 */
function createBoxStream(opts) {
  if (!opts || 'object' !== typeof opts) {
    throw new TypeError('crypto.box: Expecting object.')
  }

  // create new reference
  /* eslint-disable-next-line no-param-reassign */
  opts = { ...opts }

  if (false === isBuffer(opts.nonce)) {
    throw new TypeError('crypto.createBoxStream: Expecting nonce.')
  }

  return through(transform)

  function transform(buffer, enc, done) {
    const chunks = split(buffer, kBoxBufferSize)

    for (const chunk of chunks) {
      const offset = kBoxHeaderSize
      const boxed = box(chunk, opts)
      const nonce = increment(copy(boxed.nonce))
      const head = boxed.slice(0, offset)
      const body = boxed.slice(offset)

      this.push(head)
      this.push(body)

      Object.assign(opts, { nonce })
    }

    done(null)
  }
}

function copy(x) {
  const y = Buffer.allocUnsafe(x.length)
  x.copy(y, 0, 0, x.length)
  return y
}

module.exports = {
  createBoxStream,
  box,
}
